tests/test_metax_api.py

```diff
@@ -0,0 +1,124 @@
+     assert result == "1f32f478-8e7e-4d72-9638-d29a4f1430aa"
+
+@pytest.fixture
+def mock_post_put_response_json():
+    return '{"id": "000-000-000"}'
+
+@pytest.fixture
+def mock_requests_post(mock_post_put_response_json, metax_base_url):
+    with requests_mock.Mocker() as mocker:
+        mocker.post(f"{metax_base_url}/datasets", text=mock_post_put_response_json)
+        yield mocker
+
+def test_create_dataset_successful(mock_requests_post, caplog):
+    metadata_dict = {"persistent_identifier": "urn.fi/urn:nbn:fi:lb-201603170300", "title": {"en": "The Corpus"}, "description": {"en": "A large corpus"}, "modified": "2016-03-17T00:00:00.000000Z", "issued": "2016-03-17T00:00:00.000000Z", "access_rights": {"license": [{"url": "http://uri.suomi.fi/codelist/fairdata/license/code/undernegotiation"}], "access_type": {"url": "http://uri.suomi.fi/codelist/fairdata/access_type/code/restricted"}}}
+
+    with caplog.at_level(logging.INFO):
+        dataset_id = metax_api.create_dataset(metadata_dict)
+
+    assert dataset_id == "000-000-000"
+    assert "Created dataset" in caplog.text
+
+def test_create_dataset_failed(mock_requests_post, caplog, metax_base_url):
+    mock_requests_post.post(f"{metax_base_url}/datasets", status_code=400)
+
+    metadata_dict = {"persistent_identifier": "urn.fi/urn:nbn:fi:lb-201603170300", "title": {"en": "The Corpus"}, "description": {"en": "A large corpus"}, "modified": "2016-03-17T00:00:00.000000Z", "issued": "2016-03-17T00:00:00.000000Z", "access_rights": {"license": [{"url": "http://uri.suomi.fi/codelist/fairdata/license/code/undernegotiation"}], "access_type": {"orl": "http://uri.suomi.fi/codelist/fairdata/access_type/code/restricted"}}}
+
+    with pytest.raises(requests.exceptions.HTTPError), caplog.at_level(logging.ERROR):
+        metax_api.create_dataset(metadata_dict)
+
+    assert "Failed to create dataset" in caplog.text
+
+@pytest.fixture
+def mock_requests_put(mock_post_put_response_json, metax_base_url):
+    metax_dataset_id = "000-000-000"
+    with requests_mock.Mocker() as mocker:
+        mocker.put(f"{metax_base_url}/datasets/{metax_dataset_id}", text=mock_post_put_response_json)
+        yield mocker
+
+def test_update_dataset_successful(mock_requests_put, caplog):
+    metax_dataset_id = "000-000-000"
+    sample_dict = {"persistent_identifier": "urn.fi/urn:nbn:fi:lb-201603170300", "title": {"en": "The Corpus"}, "description": {"en": "A large corpus"}, "modified": "2016-03-17T00:00:00.000000Z", "issued": "2016-03-17T00:00:00.000000Z", "access_rights": {"license": [{"url": "http://uri.suomi.fi/codelist/fairdata/license/code/undernegotiation"}], "access_type": {"url": "http://uri.suomi.fi/codelist/fairdata/access_type/code/restricted"}}}
+
+    with caplog.at_level(logging.INFO):
+        dataset_id = metax_api.update_dataset(metax_dataset_id, sample_dict)
+
+    assert dataset_id == "000-000-000"
+    assert "Updated dataset" in caplog.text
```

Comment:
I tried to apply these suggested tests for all API calls. Could be they are still not right but really hoping they are!

Reply:
In most if not all of these cases we can simplify the assertions by just checking that the mocker has been called, this type of pattern:
```
(.venv) ajarven@mustavaris:~/tmp/mockerexample$ cat program.py 
import requests


def fetch_cat_fact():
    response = requests.get("https://catfact.ninja/fact")
    return response.json()["fact"]


if __name__ == "__main__":
    print(fetch_cat_fact())
```
```
(.venv) ajarven@mustavaris:~/tmp/mockerexample$ cat test_program.py 
import pytest
import requests_mock

from program import fetch_cat_fact


@pytest.fixture
def mock_cat_api():
    with requests_mock.Mocker() as mocker:
        mocker.get(
            "https://catfact.ninja/fact",
            json={"fact": "cats have whiskers", "length": 123},
        )
        yield mocker


def test_fetch_cat_fact(mock_cat_api):
    assert fetch_cat_fact() == "cats have whiskers"
    assert mock_cat_api.called
    assert mock_cat_api.called_once
    assert mock_cat_api.call_count == 1
```
(except we of course don't need all the three asserts, I just wanted to include the most common flavours in the example)

Because we already mock a specific URL, the test will fail if the code queries the wrong endpoint. E.g. if I change the cat fact URL from the example, I get
```
E       requests_mock.exceptions.NoMockAddress: No mock address: GET https://catfact.ninja/fact/or/something
```

When testing queries with payload that might or might not be sent correctly, we do have use for `request_history` though: with it we can access the payload of the request and check that there are as many metadata records as we expect and/or their content is right.